apis/python/src/tiledb/vector_search/object_readers/directory_reader.py

```diff
@@ -8,6 +18,223 @@
 # from tiledb.vector_search.object_readers import ObjectPartition, ObjectReader
 
 
+# Util functions for file matching and recursive listing
+def match_uri(
+    search_uri: str,
+    file_uri: str,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+) -> bool:
+    search_uri = search_uri.rstrip("/") + "/"
+    file_split = file_uri.split(search_uri)
+    file_name = "/" if len(file_split) < 2 else file_split[1]
+    file_path = Path(file_name)
+    if not file_path.match(include):
+        return False
+    if any(file_path.match(e) for e in exclude):
+        return False
+    if suffixes:
+        if file_path.suffix not in suffixes:
+            return False
+        else:
+            return True
+    return True
+
+
+def find_uris_vfs(
+    search_uri: str,
+    config: Optional[Mapping[str, Any]] = None,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+) -> Sequence[str]:
+    with tiledb.scope_ctx(config):
```

Comment:
I believe two contexts are being created here and two lines below. Perhaps it'd be better to create one ctx at the top of the stack and pass that to all functions. Generally we want to avoid repeated context creation as each context contains its own threadpools and other resources.